Feat/#22 고양이 정보 crud 작성

heyoni opened the pull request:
## Definition
- #22

heyoni committed:
fix: 로그인 로직 수정, restapi 적용, sub 변경

- 로그인 로직 수정(JWTError 추가)
- 로그인 후 JWT token sub값을 기존 id값에서 user_id(아이디)값으로 변경
- restapi 적용(슬래시 제거)

heyoni committed:
feat: cat 기능 작성, 테스트 코드 추가

- cat 모델 추가 및 service/repository 패턴 적용
- Repository 상속 모델을 따로 정의함
- foreign key 적용 후 충돌이 있어서 alembic 초기화
- cat 기능 테스트 코드 추가
- login 로직 변경 후(id -> user_id) UserService 및 테스트 코드 수정
- test 코드에서 get 요청 시 값이 없을 경우 try~except로 예외 처리 적용

f-lab-zion commented on api/deps.py:
```diff
@@ -1,19 +1,37 @@
-    if not decoded_token["sub"]:
-        raise HTTPException(status_code=404, detail="user의 id값을 찾을 수 없습니다.")
-    id = int(decoded_token["sub"])
+# def get_current_user_id(token: str = Depends(oauth2_scheme)):
+#     decoded_token = jwt.decode(token, settings.SECRET_KEY, algorithms=["HS256"])
+#     if not decoded_token["sub"]:
+#         raise HTTPException(status_code=404, detail="user의 id값을 찾을 수 없습니다.")
+#     id = int(decoded_token["sub"])
```
코드를 주석처리 하는 것은 지양해야 합니다
나중에 참고할 필요가 있는 코드면 history에서 조회할 수 도 있는데 굳이 남겨둘 필요가 없죠

f-lab-zion commented on api/deps.py:
```diff
@@ -1,19 +1,37 @@
+        user_id: str = payload.get("sub")
+        if user_id is None:
+            raise credentials_exception
+        token_data = TokenData(user_id=user_id)
+    except JWTError:
+        raise credentials_exception
+    return token_data.user_id
```
TokenData를 쓸 필요가 없어보입니다

f-lab-zion commented on repositories/UserRepository.py:
```diff
@@ -70,9 +61,10 @@ def create(self, user_new: User):
-        user = self._user[id - 1]
-        if not user:
-            return None
+        try:
+            user = self._user[id - 1]
+        except IndexError:
+            user = None
```
Exception 으로 처리하는 방법 외에 `self._user` 의 length 를 이용하여 미리 체크하는 방식도 있습니다

heyoni replied:
try except를 사용하는 것 보다는 raise exception이 더 효율적인가요?

f-lab-zion replied:
이 경우 index인 `id` 가 0보다 작거나 `len(self._user)` 보다 크다면 return None 하고 그렇지 않으면 `return self._user[id - 1]`으로 처리하면 의미가 더 명확할 수 있다는 의미입니다. 물론 `IndexError` 가 해당 케이스를 exception으로 잡아주는 것은 자명합니다만 코드의 가독성과 협업 측면에서는 의미를 보다 명확하게 드러나는 코드가 좋아보입니다. 이건 코드 스타일에 대한 문제라서 어느 것이 옳다 그르다를 말하기 어렵습니다만 하나의 의견으로 들어주세요.

f-lab-zion commented on repositories/CatRepository.py:
```diff
@@ -0,0 +1,94 @@
+            return None
+        return cat
+
+    def get_user_by_user_id(self, current_user_id: str) -> User:
+        user = self.db.query(User).filter(User.user_id == current_user_id).first()
+        if not user:
+            return None
+        return user
```
UserRepository의 함수와 역할이 동일해 보입니다

f-lab-zion commented on services/CatServices.py:
```diff
@@ -0,0 +1,89 @@
+        return cat
+
+    def create(self, cat_body: CatCreate, current_user_id: str) -> Cat:
+        user = self.catRepository.get_user_by_user_id(current_user_id)
```
CatService 가 꼭 CatRepository만 접근할 필요는 없습니다
경우에 따라 UserRepository에 접근해도 됩니다

f-lab-zion commented on services/CatServices.py:
```diff
@@ -0,0 +1,89 @@
+        if cat.user.id != user.id:
+            raise HTTPException(status_code=401, detail="user가 다릅니다. 권한이 없습니다.")
+
+        return self.catRepository.delete(id)
```
대표 고양이가 지워질 경우 남은 고양이 중에 대표 고양이를 결정해야 하지 않나요?

heyoni replied:
고양이가 한마리만 있을 경우도 있기 때문에 대표고양이 설정은 유저 선택에 맡길 예정입니다!
(대표고양이가 설정이 안되어있으면 UI로 '대표고양이 설정이 필요합니다'라는 안내가 나갈 예정입니다.)

f-lab-zion reviewed (changes requested):
👍 수고하셨습니다 상세 내용은 코멘트를 참고해주세요

heyoni committed:
feat: 코드 리뷰 반영

- 주석 삭제
- TokenData 삭제
- Repository에서 try-except 대신 len을 이용하여 return None 처리
- CatRepository에서 UserRepository를 사용하여 중복 코드 제거

f-lab-zion reviewed (approved):
👍 좋아보입니다
